#!/usr/bin/env python3

# Base paths
BASE_DIR = "/beegfs/scratch/ric.broccoli/kubacki.michal/SRF_MeCP2_CUTandTAG/Cross_final"

# Data paths
DATA_DIR = f"{BASE_DIR}/data"
GENOME_FASTA = "/beegfs/scratch/ric.broccoli/kubacki.michal/SRF_MeCP2_CUTandTAG/DATA/mm10.fa"
GTF_FILE = "/beegfs/scratch/ric.broccoli/kubacki.michal/SRF_MeCP2_CUTandTAG/DATA/gencode.vM25.annotation.gtf"

# Input data paths
MEDIP_DIR = "/beegfs/scratch/ric.broccoli/kubacki.michal/DATA/MECP2/MEDIP/output_done/bigwig"
SMARCB1_DIR = "/beegfs/scratch/ric.broccoli/kubacki.michal/SRF_MeCP2_SMARCB1/results/bigwig"

# Output paths
RESULTS_DIR = f"{BASE_DIR}/results"
METHYLATION_ENRICHMENT_DIR = f"{RESULTS_DIR}/1analyze_methylation_enrichment" 